frontend/py_readline.py

```diff
@@ -140,7 +140,34 @@ def print_shell_cmd_map(self):
 
     def unbind_keyseq(self, keyseq):
         # type: (str) -> None
-        line_input.unbind_keyseq(keyseq)
+        line_input.unbind_keyseq(keyseq)    
+
+    def bind_shell_command(self, bindseq):
+        # type: (str) -> None
+        cmdseq_split = bindseq.strip().split(":", 1)
+        if len(cmdseq_split) != 2:
+            raise ValueError("%s: missing colon separator" % bindseq)
+
+        # Below checks prevent need to do so in C, but also ensure rl_generic_bind
+        # will not try to incorrectly xfree `cmd`/`data`, which doesn't belong to it
+        keyseq = cmdseq_split[0].rstrip()
+        if len(keyseq) <= 2:
+            raise ValueError("%s: empty/invalid key sequence" % keyseq)
+        if keyseq[0] != '"' or keyseq[-1] != '"':
+            raise ValueError(
+                "%s: missing double-quotes around the key sequence" % keyseq)
+        keyseq = keyseq[1:-1]
+
+        cmd = cmdseq_split[1]
+        line_input.bind_shell_command(keyseq, cmd)
+
+    def set_bind_shell_command_hook(self, hook):
+        # type: (Callable[[str, str, int], Tuple[int, str, str]]) -> None
+
+        if hook is None:
```

Comment:
Should this be an assert like `assert hook is not None, "..."` ?

Reply:
Sure, np. I loosely based it off `set_completion_display_matches_hook`, which does something similar, but unlike completion, the bind-x callback should always be present (tho we may want to defer setting it until actually used)

Reply:
Done